core/lib/src/response/responder.rs

```diff
@@ -217,34 +219,156 @@ impl<'r> Responder<'r> for String {
+                            .expect("Attempted to seek to the start of the requested range, but failed.");
+                        (start, end, size)
+                    },
+                    ByteRangeSpec::AllFrom(mut start) => {
+                        let size = body.seek(io::SeekFrom::End(0))
+                            .expect("Attempted to retrieve size by seeking, but failed.");
+                        if start > size {
+                            start = size;
+                        }
+                        body.seek(io::SeekFrom::Start(start))
+                            .expect("Attempted to seek to the start of the requested range, but failed.");
+                        (start, size, size)
+                    },
+                    ByteRangeSpec::Last(len) => {
+                        let size = body.seek(io::SeekFrom::End(0))
+                            .expect("Attempted to retrieve size by seeking, but failed.");
+                        // we could seek to SeekFrom::End(-len), but if we reach a value < 0, that is an error.
+                        // but the RFC reads:
+                        //      If the selected representation is shorter than the specified
+                        //      suffix-length, the entire representation is used.
+                        let start = cmp::max(size - len, 0);
+                        body.seek(io::SeekFrom::Start(start))
+                            .expect("Attempted to seek to the start of the requested range, but failed.");
+                        (start, size, size)
+                    }
+                };
+
+                return Ok(
+                    Response::build()
+                        .status(Status::PartialContent)
+                        .header(AcceptRanges(vec![RangeUnit::Bytes]))
+                        .header(ContentRange(ContentRangeSpec::Bytes {
+                            range: Some((start, end)),
+                            instance_length: Some(total),
+                        }))
+                        .raw_body(Body::Sized(body, end - start))
+                        .finalize()
+                )
+            }
+            // A server MAY ignore the Range header field.
+        },
+        // An origin server MUST ignore a Range header field that contains a
+        // range unit it does not understand.
+        Range::Unregistered(_, _) => {},
+    };
+
+    Err(body)
+}
+
 /// Returns a response with Content-Type `application/octet-stream` and a
 /// fixed-size body containing the data in `self`. Always returns `Ok`.
 impl<'r> Responder<'r> for &'r [u8] {
-    fn respond_to(self, _: &Request) -> response::Result<'r> {
+    fn respond_to(self, req: &Request) -> response::Result<'r> {
+        let mut body = Cursor::new(self);
+
+        //  A server MUST ignore a Range header field received with a request method other than GET.
+        if req.method() == Method::Get {
```

Comment:
I think most of this code here could be broken out into a single function, maybe called `try_range_response`.

Reply:
I have moved more of into `try_range_response`.